src/OpenTelemetry/Trace/Builder/TracerProviderBuilderExtensions.cs

```diff
@@ -96,6 +96,37 @@ public static TracerProviderBuilder SetSampler<T>(this TracerProviderBuilder tra
             return tracerProviderBuilder;
         }
 
+        /// <summary>
+        /// Sets the sampler on the provider.
+        /// </summary>
+        /// <remarks>
+        /// Note: The type specified by <typeparamref name="T"/> will be
+        /// registered as a singleton service into application services.
+        /// </remarks>
+        /// <typeparam name="T">Sampler type.</typeparam>
+        /// <param name="tracerProviderBuilder"><see cref="TracerProviderBuilder"/>.</param>
+        /// <param name="implementationFactory">The factory that creates the service.</param>
+        /// <returns>The supplied <see cref="TracerProviderBuilder"/> for chaining.</returns>
+        public static TracerProviderBuilder SetSampler<T>(
+            this TracerProviderBuilder tracerProviderBuilder,
+            Func<IServiceProvider, T> implementationFactory)
```

Comment:
With the addition of overloads accepting the `implementationFactory` are there use cases that come to mind where the more general approach of using `ConfigureBuilder` is still useful?

Reply:
I guess the answer is for enabling instrumentation authors to instantiate their instrumentation using services registered with the service provider.

Reply:
@noahfalk asked me the same question. I didn't think too hard on it because `ConfigureBuilder` essentially has already shipped via the "IDeferred[Signal]ProviderBuilder" interfaces we have in API. All that is happening in 1.4.0 is it is getting a new name (it used to be "Configue") so it sits more nicely with "ConfigureServices". We could remove the extension with the new name and force people to access it by casting to the interface but that makes our exporter boilerplate even uglier/more mysterious 🤣 Thoughts?

Reply:
Do we have an AddInstrumentation<T>(Func<IServiceProvider, T> implementationFactory) overload or could we have it if one doesn't currently exist? I assume that satisfies the instrumentation case?

Reply:
@noahfalk @alanwest 

OK I updated this PR with removal of `ConfigureBuilder` from the public API. I actually really like the result. All the exporter extensions look much cleaner now. I had to use the "IDeferred[Signal]ProviderBuilder" cast trick in a couple spots in instrumentation because they were doing some really strange things. But all of the sane/expected use cases were covered. Check out the latest diff and let me know what you guys think.

> Do we have an AddInstrumentation(Func<IServiceProvider, T> implementationFactory) overload

[Yes that one we do have](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Extensions.DependencyInjection/Trace/OpenTelemetryDependencyInjectionTracerProviderBuilderExtensions.cs#L78) (also for metrics). Added earlier in 1.4.0 development 😄

Reply:
Great, I'm glad it worked out well!

Reply:
Yeah, that looks good to me :)